### ISSUE TITLE
Allow passing options to formatters

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [palantir/tslint#4281]
Related: #4223 for whether prose (soon, verbose) should output relative or full file paths.

It could be useful to be able to pass options to formatters. For example, if we default to either relative or full file paths, some users may want the other behavior.

Instead of just taking in `failures` and `fixes` sequentially:

```typescript
 interface IFormatter {
    format(failures: RuleFailure[], fixes?: RuleFailure[]): string;
}
```

It'd be nice to pass formatters an object with those fields and an `options`. Something like:

```typescript
interface IFormatOptions<TOptions> {
    failures: RuleFailure[];
    fixes?: RuleFailure[];
    options: TOptions;
}

interface IFormatter<TOptions = {}> {
    format(formatOptions: IFormatOptions<TOptions>): string;
}
```

...but to maintain backwards compatibility with older TSLint versions, we'd want to still pass in a `RuleFailure[]` and `fixes?: RuleFailure[]`... so:

```typescript
type IFormatOptions<TOptions> = RuleFailure[] & {
    failures: RuleFailure[];
    fixes?: RuleFailure[];
    options: TOptions;
};

interface IFormatter<TOptions = {}> {
    format(formatOptions: IFormatOptions<TOptions>, fixes?: RuleFailure[]): string;
}
```

...where the first object is an array with the `failures`, `fixes`, and `options` members added to it. The second parameter, `fixes`, would be deprecated, and we would want to eventually get rid of it.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for major new changes or features. See #4534. 😱

If you'd like to see this change implemented, you have two choices:
* **Recommended:** Check if this is available in [ESLint](https://eslint.org) + [typescript-eslint](https://typescript-eslint.io) ✅
* _Not Recommended: Fork TSLint locally_ 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_
